slang_frontend.cc

```diff
@@ -997,14 +1003,39 @@ struct ProceduralVisitor : public ast::ASTVisitor<ProceduralVisitor, true, false
 			if (!cv.as_const().as_bool())
 				break;
 
+			RTLIL::SigSpec disable;
+
+			if (disableLoop_outs.empty()) {
+				disable = RTLIL::S0;
+			} else {
+				disable = disableLoop_outs.back();
+			}
+
+			// add next disable wire
+			RTLIL::Wire *breakNextLoop = netlist.canvas->addWire(NEW_ID_SUFFIX("breakNextLoop"), 1);
+			RTLIL::Wire *breakNextLoopOred = netlist.canvas->addWire(NEW_ID_SUFFIX("breakNextLoopOred"), 1);
+			do_simple_assign(stmt.sourceRange.start(), breakNextLoop, RTLIL::S0, false);
+			netlist.canvas->addOr(NEW_ID_SUFFIX("breakNextLoopOr"), breakNextLoop, disable, breakNextLoopOred, false);
```

Comment:
Please add `RTLILBuilder::LogicOr()`, and do `breakNextLoopOred = netlist.LogicOr(breakNextLoop, disable);` here.

(Right now `RTLILBuilder` is just a thin wrapper which does constant folding, which is not useful here, but in the future it may do more, so we should be using it consistently.)

Reply:
Having thought about this more, instead of ORing the disable signals explicitly, and using non-blocking assignments, we want to rely on the handling of blocking assignments and do the following:

 * in handling of a loop, create a single disable wire with the attribute `$nonstatic` (see `create_local` which does the same for user variables with automatic lifetime)
 * initialize the disable wire to zero by a blocking assignment to it
 * in the handling of a break statement, use a blocking statement to set the disable signal to one
 * at the start of each iteration, pass the disable wire to `substitute_rvalue` to evaluate it at that point within the procedure, and use the signal that's returned to drive the switch that disables the loop body
 
 Note that the wire itself which is created for variables with automatic lifetime isn't ever driven, it's only a placeholder to have a key into the maps inside `VariableState` to track the value that's been assigned to them (since the automatic variables cannot be referred to outside of a procedure, we need to track the value that's been assigned to them, but we don't need to create a driver for them)